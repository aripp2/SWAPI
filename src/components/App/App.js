import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import './App.scss';
import UserProfile from '../UserProfile/UserProfile';
import MoviesContainer from '../MoviesContainer/MoviesContainer';
import CharactersContainer from '../CharactersContainer/CharactersContainer';
import UserForm from '../UserForm/UserForm';
import { getMovies, fetchCharacters } from '../../apiCalls';


class App extends Component {
  constructor() {
    super();
    this.state = {
      user: {},
      movies: [],
      selectedCharacters: [],
      favorites: [],
      haveUser: false,
      isLoading: true,
      error: ''
    }
  }

  componentDidMount() {
    getMovies()
      .then(allFilms => this.setState({ movies: allFilms, isLoading: false }))
      .catch(error => this.setState({ error: 'Something is wrong in the universe'}))
    }

  submitUser = (user) => {
    this.setState({user, haveUser: true })
  }
  
  getCharacters = (charactersUrls) => {
    this.setState({isLoading: true})
    fetchCharacters(charactersUrls)
      .then(data => this.setState({ selectedCharacters: data, isLoading: false }))
      .catch(error => this.setState({ error: 'Something is wrong in the universe'}))
  }

  addFavorite = (character) => {
    const { favorites } = this.state
    const updatedFavs = [...favorites, character]
      this.setState({ favorites: updatedFavs })
  }

  removeFavorite = (character) => {
    const { favorites } = this.state
    const updatedFavs = favorites.filter(fav => fav.name !== character.name)
    this.setState({ favorites: updatedFavs })
  }

  updateFavs = (character) => {
    const { favorites } = this.state
    const names = favorites.map(fav => fav.name)
     names.includes(character.name) ? this.removeFavorite(character) : this.addFavorite(character)
  }

  render() {
    const { user, movies, selectedCharacters, favorites, haveUser, isLoading, error } = this.state;
    
    return (
      <div>
        <Route exact path='/' render={() => 
          <UserForm
            submitUser={this.submitUser} />} />
        <main>
          {haveUser && <UserProfile {...user} favorites={favorites}/>}
          {error && <h2>{error}</h2>}
          <Route exact path='/movies' render={() => 
            <MoviesContainer 
              movies={movies} 
              getCharacters={this.getCharacters} />} />
          <Route exact path='/movies/:id' render={({ match }) => {
            const selectedMovie = movies.find(movie => movie.episode_id === parseInt(match.params.id))
            return (
            <CharactersContainer 
              characters={selectedCharacters}
              updateFavs={this.updateFavs} 
              movieInfo={selectedMovie} 
              isLoading={isLoading} />)}} />
          <Route exact path='/favorites' render={() => 
            <CharactersContainer 
              characters={favorites}
              updateFavs={this.updateFavs}
                />} />
        </main> 
      </div>
    )
  }
};

export default App;



