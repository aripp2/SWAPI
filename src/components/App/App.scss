@import url('https://fonts.googleapis.com/css?family=Luckiest+Guy&display=swap');

$title-font: 'Luckiest Guy', cursive;


h1{
  font-family: $title-font;
  letter-spacing: 2px;
}

.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}
