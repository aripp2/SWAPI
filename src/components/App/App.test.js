import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme'; 
import App from './App';


describe('App', () => {
  let wrapper; 

  beforeEach(() => {
    wrapper = shallow(<App />)
  });

  it('should match the Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should update state when submitUser is called' () => {
    
  })

}); 


