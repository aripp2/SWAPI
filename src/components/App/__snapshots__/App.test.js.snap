// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the Snapshot 1`] = `
<div>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <main>
    <h1>
      STAR WARS
    </h1>
    <Route
      exact={true}
      path="/movies"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/movies/:id"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/favorites"
      render={[Function]}
    />
  </main>
</div>
`;
